lib/fdo.c

```diff
@@ -695,6 +699,111 @@ fdo_sdk_status fdo_sdk_init(fdo_sdk_errorCB error_handling_callback,
 	return FDO_SUCCESS;
 }
 
+/**
+ * Parse the manufacturer network address in the given buffer, and extract and save
+ * the TLS/IP/DNS/Port values.
+ *
+ * @param buffer Buffer containing the network address
+ * @param tls Output flag describibg whether HTTP (false) or HTTPS (true) is used
+ * @param mfg_ip Output structure to store IP. Null is returned if no IP is present
+ * @param mfg_dns Output pre-allocated buffer to store DNS
+ * @param mfg_dns_sz Size of the DNS buffer (minimum 100)
+ * @param mfg_port Output variable to store port
+ *
+ * Return true if parse was successful, false otherwise.
+ */
+bool parse_manufacturer_address(char *buffer, bool *tls,
+	fdo_ip_address_t **mfg_ip, char *mfg_dns, size_t mfg_dns_sz,
+	int *mfg_port) {
+
+	char transport_prot[6] = {0};
+	char port[6] = {0};
+	size_t dns_sz = 100;
+	int result = 0;
+	char *eptr = NULL;
+	const char transport_http[5] = "http";
+	const char transport_https[6] = "https";
+
+	if (!buffer || !tls || !mfg_ip || !mfg_dns ||
+		mfg_dns_sz < dns_sz || !mfg_port) {
+		LOG(LOG_ERROR, "Invalid params\n");
+		return false;
+	}
+
+	// the expected format is '{http/https}://{IP/DNS}:port', port is optional
+	result = sscanf(buffer, "%5[^:]://%99[^:]:%5s/", transport_prot, mfg_dns, port);
+
+	// if all 3 are updated, wer're all good,
+	// but if only 2 of these are updated, verify each and decide
+	if (result != 2 && result != 3) {
+		LOG(LOG_ERROR, "Failed to parse manufacturer address\n");
+		goto end;
+	}
+
+	// parse transport protocol. check for 'http' first, then 'https'
+	*tls = false;
+	if (memcmp_s(transport_prot, sizeof(transport_prot), transport_http,
+			sizeof(transport_http), &result) != 0) {
+		LOG(LOG_ERROR, "Failed to compare transport protocol\n");
+		goto end;
+	}
+	if (0 != result) {
+		if (memcmp_s(transport_prot, sizeof(transport_prot), transport_https,
+			sizeof(transport_https), &result) != 0) {
+			LOG(LOG_ERROR, "Failed to compare transport protocol\n");
+			goto end;
+		}
+		if (0 == result) {
+			*tls = true;
+			LOG(LOG_DEBUG, "Manufacturer Transport protocol: HTTPS\n");
+		} else {
+			LOG(LOG_ERROR, "Invalid Manufacturer Transport protocol specified.\n");
+			goto end;
+		}
+	} else {
+		LOG(LOG_DEBUG, "Manufacturer Transport protocol: HTTP\n");
+	}
+
+	// parse IP/DNS, check for IP first, if it fails, treat it as DNS
+	*mfg_ip = fdo_ipaddress_alloc();
+	if (!mfg_ip) {
+		LOG(LOG_ERROR, "Failed to alloc memory\n");
+		ERROR();
+		goto end;
+	}
+	result = fdo_printable_to_net(mfg_dns, (*mfg_ip)->addr);
+	if (result > 0) {
```

Comment:
mfg_ip is freed in one case and not the other - is there a memory leak here?

Reply:
The input can contain either one of IP or DNS. 
If IP is specified, then we don't free the IP and use it later.
If DNS is specified, then the IP is freed and returned as NULL entity.

Further, the method is responsible for memory allocation and return NULL (DNS to be used) or non-NULL (IP to be used). If non-NULL, then it is freed in caller STATE_DI().

Reply:
@prateekcr It would be good to add a comment in line 768 or 708 explaining how mfg_ip is freed in different conditions. That would help someone in future when they look at this.